factory/FLINTconvert.cc

```diff
@@ -112,16 +112,19 @@ void fq_nmod_set_nmod_poly(fq_nmod_t a, const nmod_poly_t b,
 
 #include "FLINTconvert.h"
 
+// TODO!! decide if the input to convertCF2Fmpz is expected to be:
+//  initialized, or
+//  uninitialized
 void convertCF2Fmpz (fmpz_t result, const CanonicalForm& f)
 {
   if (f.isImm())
-    *result=f.intval();
+    *result=f.intval();                     // assumes uninitialized
```

Comment:
@hannes14 This is actually an important TODO. The rest of the convertCF2flint functions seem to assume that their inputs are uninitialized (which is not nec. a bad thing), but the majority of the use cases of convertCD2fmpz seems to assume that `result` is initialized. This line `*result=f.intval();` has the potential to leak a lot of integers, and the commented out line `fmpz_set_si (result, f.intval())` would be the correct way if `result` is initialized.

Reply:
I will change it to "result is uninitialized" here and change all uses case which assumes the other way.

Reply:
I would advocate having both since there are a bunch of functions already written assuming initialized input.
```
// as with the other convert functions, result comes in uninitialized
void convertCF2Fmpz (fmpz_t result, const CanonicalForm& f)
{
  if (f.isImm())
    *result=f.intval();
  else
  {
    // promote + demote_val avoids copying the data twice
    fmpz_init(result);
    f.mpzval(_fmpz_promote(result));
    _fmpz_demote_val(result);
  }
}

// special version assuming result is initialized
void convertCF2initFmpz (fmpz_t result, const CanonicalForm& f)
{
  if (f.isImm())
    fmpz_set_si (result, f.intval());
  else
  {
    f.mpzval(_fmpz_promote(result));
    _fmpz_demote_val(result);
  }
}
```

and then the use is

```
void convertFacCF2Fmpz_poly_t (fmpz_poly_t result, const CanonicalForm& f)
{
  fmpz_poly_init2 (result, degree (f)+1);
  _fmpz_poly_set_length(result, degree(f)+1);
  for (CFIterator i= f; i.hasTerms(); i++)
    convertCF2initFmpz (fmpz_poly_get_coeff_ptr(result, i.exp()), i.coeff());
}
```

Reply:
I think there should be two functions since a lot of code is currently written
using initialized input:
```
// as with the other convert functions, result comes in uninitialized
void convertCF2Fmpz (fmpz_t result, const CanonicalForm& f)
{
  if (f.isImm())
    *result=f.intval();
  else
  {
    mpz_t gmp_val;
    f.mpzval(gmp_val);

    // promote + demote_val avoids copying the data twice
    fmpz_init(result);
    mpz_swap(gmp_val, _fmpz_promote(result));
    _fmpz_demote_val(result);

    mpz_clear (gmp_val);
  }
}

// special version assuming result is already initialized
void convertCF2initFmpz (fmpz_t result, const CanonicalForm& f)
{
  if (f.isImm())
    fmpz_set_si (result, f.intval());
  else
  {
    mpz_t gmp_val;
    f.mpzval(gmp_val);

    mpz_swap(gmp_val, _fmpz_promote(result));
    _fmpz_demote_val(result);

    mpz_clear (gmp_val);
  }
}
```

and then since fmpz_poly_init2 creates initialized coeffs, the use is

```
void convertFacCF2Fmpz_poly_t (fmpz_poly_t result, const CanonicalForm& f)
{
  fmpz_poly_init2 (result, degree (f)+1);
  _fmpz_poly_set_length(result, degree(f)+1);
  for (CFIterator i= f; i.hasTerms(); i++)
    convertCF2initFmpz (fmpz_poly_get_coeff_ptr(result, i.exp()), i.coeff());
}
```